### ISSUE TITLE
TPM2 measurement of mount units and unlocked LUKS volumes

### ISSUE THREAD

**Issue by berrange** [systemd/systemd#24503]
### Component

systemctl, systemd-cryptsetup, systemd-gpt-auto-generator

### Is your feature request related to a problem? Please describe

With a combination of sd-boot, SecureBoot, and a unified kernel image, it is possible construct a virtual machine disk image that has a trusted boot path, eliminating the traditional hole that Linux distros have with a locally created initrd + cmdline not being measured into the TPM.  To avoid need to hardcode filesystem / LUKS volume UUIDs in the kernel cmdline, the image would typically be built in compliance with the systemd discoverable partitions specification, allowing the distro vendor to build a single unified kernel image, for use in many different scenarios.

With confidential computing it is necessary to eliminate trust of the host compute administrator & software stack, and approaches to this share many of the concepts/features related to TPMs/SecureBoot. Taking AMD SEV-SNP as a example, assuming the initial virtual machine firmware and unified kernel are verified with a remote attestation agent, a key can be released to unlock the state of a vTPM running in VMPL 0 (the guest OS could be VMPL 1/2/3). This trusted vTPM (as opposed to an untrusted vTPM running in the host) can now be used to unseal the key needed to access the LUKS encrypted root filesystem image in the normal manner, exposing the discoverable partitions to systemd-gpt-auto-generator.

There is a risk though that the malicious/compromised host OS may have substituted the intended root filesystem with one of their choosing, but with the same /boot & /boot/efi contents, thus preserving the signed & measured unified kernel image. systemd in the signed initrd will happily auto-discover and mount partitions from the rogue root filesystem.

A variant on this theme would be for the normal root filesystem to have been provided, but the malicious host admin provides a 2nd virtual disk containing a replacement /home or /usr that would be discovered and overmounts the /home or /usr present in the root filesystem.

There needs to be a way for the tenant of the guest VM to verify that the filesystems that are mounted are exactly the ones they expected from their VM disk image.

A second aspect is that in confidential computing, since the host OS/admin is considered untrusted, by implication the virtual keyboard, virtual serial port and virtual VGA device are also untrusted. As such a guest admin should never attempt to enter any LUKS passphrase interactively during early boot, as those keyboard input events would be visible to the host OS.  Nor must any virtual USB security key be used. Only the vTPM is trust worthy, since its state is running in the guest in VMPL 0 and only unlocked once the VM firmware is attested. 

In addition to the trusted vTPM running in VMPL 0, the host admin might have exposed a virtual TPM running in the host OS which should not be trusted.

Thus there also needs to be a way for the tenant of the guest VM to verify how the LUKS volume was unlocked - passphrase vs TPM vs USB security key vs something else. The source of the credential type needs to be identifiable to, eg which of the many vTPMs were used.

So overall the key threats we're aiming to detect are

  - Host admin substituting the root disk with a non-encrypted one
  - Host admin substituting the root disk with an encrypted one of their choice, and secretly unlocking with a passphrase
  - Host admin adding a second disk with extra volumes, eg over-mounting /home in the root disk, with a /home from a 2nd disk
  - Host admin substituting the root disk with an encrypted one of their choice, and a second untrusted vTPM to unlock it
  - Guest tenant compromising the confidentiality of the VM themselves by mistakenly unlocking with an interactively entered passphrase from the non-confidential keyboard

### Describe the solution you'd like

Since an AMD SEV-SNP  VM has a trustworthy vTPM available due to early attestation of the confidential VM firmware/hardware, the vTPM PCRs+eventlog are a natural choice for verifying many aspects of the guest OS boot up choices. 

If, during the process of unlocking LUKS volumes and auto-discovering partitions for .mount units, systemd were to measure key data into a TPM PCR and event log, it would be possible to remotely attest that the root filesystem is the one that was expected.

For each LUKS volume discovered and unlocked, systemd would need to measure 

 * the LUKS volume UUID
 * a string reflecting the credential type used for unlocking it. 'pkcs#11', 'fido2', 'tpm2', 'passphrase', 'recoverykey'.
 * any relevant credential type specific data (eg which tpm2 device was used)

For each partition mounted, systemd would need to measure

* the partition type GUID
* partition attribute flags (eg to identify readonly vs readwrite mounts, and other properties) 
* the device holding the partition (physical vs a device-mapper (LUKS) volume

With that information, I believe it should be possible for a remote attestation service policy to verify that the filesystems discovered and mounted came from the specific encrypted LUKS volume, unlocked via the trusted vTPM.

### Describe alternatives you've considered

The systemd unit file configs inside the signed and attested initrd could be populated to disable all use of partition auto-discovery and only allow unlocking via the specific trusted vTPM. The challenge with is that writing and maintaining such a config could be a non-trivial burden and has a high risk of inadvertent mistakes allowing an exploit of the confidential VM disk image. So even if going this route of disabling auto-discovery via unit file configs, it would still be desirable to have details of all unlocked LUKS volumes and mounted filesystems measured into the vTPM for attestation post-boot.

Another alternative is to measure the entire virtual disk contents that holds the root FS, and verify this at the same time as verifying the firmware and unified kernel image.  This is less flexible as any time the disk contents change, updated measurements needs to be made available to the attestation service. This is likely to be impractical in the general case.  Again even if the root disk is measured, there's still a chance that is was built in a way that made it susceptible to a malicious admin attaching a 2nd virtual disk to the VM with discoverable overrides for sub-mounts on /home. So again it looks like it is desirable to be able to measure any info about LUKS volumes and mounted filesystems into the vTPM.

It was considered whether cryptsetup can write the PCRs measurements  / eventlog, however, it does not appear to have sufficient information about the type of credentials systemd used to unlock the volume, as all the different systemd crypttab credential types just end up calling the same crypt_activate_by_volume_key and crypt_activate_by_passphrase APIs.

The kernel IMA logs have record of the device mapper targets loaded, but again don't appear to have information about the source of the credentials used to unlock the volume.  The kernel could also possibly be used as a source of information about mounts performed, but the information it has is not quite so easily mapped back to the expected configuration. eg a kernel IMA log might tell you the the device major,minor number, but these are dynamically generated and not easily associated with a specific GPT partition GUID which your config defined.

**Comment by poettering** [comment]
So, in my thinking there are two ways to bind a kernel/TPM combination to an OS file system.

1. Use verity, encode the root hash in the kernel cmdline embedded to a unified kernel image. This is of course incompatible with systems which deploy RPM for OS updates

2. Lock the root fs to the kernel + TPM via the stuff prepped in #24351 + #24458. Set "headless" mode to ensure that passphrase stuff cannot be used.

Both in combination (i.e. verity for /usr/ and the tpm-pcr-signed stuff for root fs) are a good choice too.

There's currently logic in place for the gpt auto logic to only look for partitions on the ESP disk. However the ESP is identified from userspace via its UUID which is untrusted of course.

I think measuring information about activated LUKS volumes and file systems makes a ton of sense. But it needs some careful implementation since just measuring keys and mountpoint info is not enough: we must do this in a specific order, otherwise the resulting PCR values will be garbage and non-deterministic.

So basically, we'd need to enforce some ordering between mounts we traditionally parallelized, so that the measurements happen in a well-defined order.

There's a TODO list item in the TODO file in git where I already suggested we should measure root fs info, but I assumed this might be mostly redundant if IMA is in the mix, in conjunction with "headless" mode, that basically disables password auth, and thuse only allows the listed unlocking which then would be TPM2. But my understanding is you want more flexibility there, and actually allow the admin to make changes to crypttab?

I presume you want to measure all mounts being established, not just the ones via the gpt-auto logic? i.e. manual and fstab stuff too?

**Comment by bluca** [comment]
> The kernel IMA logs have record of the device mapper targets loaded, but again don't appear to have information about the source of the credentials used to unlock the volume. The kernel could also possibly be used as a source of information about mounts performed, but the information it has is not quite so easily mapped back to the expected configuration. eg a kernel IMA log might tell you the the device major,minor number, but these are dynamically generated and not easily associated with a specific GPT partition GUID which your config defined.

To me it seems like the only possible way to do this correctly is via the IMA log. dm-verity already measures key metadata, and if dm-crypt doesn't then it probably should?
The point is that if you are trying to stop local modifications, simply changing systemd is not going to be enough. Anything with root can mount anything wherever they want to, manually, and they just have to not measure anything and it's all moot, no?
That's why I say that this needs to be handled in the kernel, as that's where the mounts actually happen, so you can be confident that anything that happens on the system is logged, regardless of the source.

**Comment by berrange** [comment]
> 1. Use verity, encode the root hash in the kernel cmdline embedded to a unified kernel image. This is of course incompatible with systems which deploy RPM for OS updates

Right, I should have mentioned that my starting point is that I'm trying to fit into the traditional way of building and distributing RPMs and cloud images in Fedora / RHEL as much as practical. I want to avoid a need to reinvent too much of the distro, to reduce barriers to getting something useful into hands of users. My desire is that there would be a single kernel-unified-virt RPM shipping a pre-built unified kernel, whose initrd includes drivers for KVM, HyperV, Xen, VMWare, along with a totally generic cmdline (perhaps nothing more than 'console=ttyS0 console=tty quiet rhgb'), signed by the OS vendor with their SecureBoot certs. This kernel should be usable together with shim + sd-boot, in arbitrary cloud disk images created either by the OS vendor or by end users.  Ideally the image would be bootable in any cloud scenario....  EFI, EFI + vTPM, EFI + SecureBoot, EFI + SecureBoot + vTPM, EFI + confidential virt + vTPM  ...  with the user having the option to attest the image once booted if it has a vTPM, dependant on what level of trust/confidentiality they need to prove at runtime.

> Lock the root fs to the kernel + TPM via the stuff prepped in https://github.com/systemd/systemd/pull/24351 + https://github.com/systemd/systemd/pull/24458. Set "headless" mode to ensure that passphrase stuff cannot be used.

I'm not sure I entirely understand how this all fits together to solve the problem yet. Will keep reading but if you have any more verbose guidance that'd be useful.

> I think measuring information about activated LUKS volumes and file systems makes a ton of sense. But it needs some careful implementation since just measuring keys and mountpoint info is not enough: we must do this in a specific order, otherwise the resulting PCR values will be garbage and non-deterministic.
>
> So basically, we'd need to enforce some ordering between mounts we traditionally parallelized, so that the measurements happen in a well-defined order.

Right, deterministic ordering does look important to be able to correctly interpret the mount setup.

> There's a TODO list item in the TODO file in git where I already suggested we should measure root fs info, but I assumed this might be mostly redundant if IMA is in the mix, in conjunction with "headless" mode, that basically disables password auth, and thuse only allows the listed unlocking which then would be TPM2. But my understanding is you want more flexibility there, and actually allow the admin to make changes to crypttab?

My primary focus is address the needs of confidential computing. To make confidential computing transparent & simple for users, however, I don't want to do special things just for the CC use case unless completely unavoidable. Thus my preference is to find a way to construct a cloud disk image that will work in any reasonable EFI based virtual machine.  So while with confidential computing I can assume headless and no interactive input at boot, I do want the cloud disk image to support interaction at boot for when users are not running on a confidential computing hardware. Thus the thought that rather than blocking functionality up front, we instead measure what we are doing into any TPM that exists, allowing users to attest to the state once boot is complete. This also gives detection abilities should the user make mistakes in their disk image construction/configuration, which I think will be all to common.

**Comment by poettering** [comment]
I see, so it's a lot more about having measurements of what happened than about blocking what can happen, right?

I don't really grok what "confidential computing" really means. I mean, I know some keywords, but it appears to me there's a lot of handwaving and unclear semantics... 

Is there a way to detect that I am running in a CC environment? Is there even a precise semantic definition of what this would even mean?

So I wonder how useful the measurements would actually be, if we trigger them in userspace from systemd, given that systemd is not the only userspace around and basically any userspace code can call DM ioctls and the mount() syscalls. Hence, it appears to me that what you probably want is to do this in kernel space. But there of course it's hard to know how precisely a volume was unlocked, i.e. via tpm2 or passphrase or something else. Moreover, things can get very confusing given that we are living in a fs namespaced world, and thus only mounts established in the host mount table matter, what happens in the per-service fs namespaces doesn't matter at for measurements...

One question that would need answering: if i boot up, and all is good, so you got measurements for /home/ and /srv/. And then after boot, much later some admin scrip unmounts /srv/ and mounts something else there. Should that show up in the measurements?

**Comment by bluca** [comment]
> > The point is that if you are trying to stop local modifications, simply changing systemd is not going to be enough. Anything with root can mount anything wherever they want to, manually, and they just have to not measure anything and it's all moot, no?
> 
> I am **not** trying to stop or detect local modifications by the guest admin.
> 
> The threat is a malicious host OS that compromises the guest disk image in some manner. Thus the goal is establish trust of the disk mount setup after boot, to prove that the guest disk image the guest admin uploaded to the cloud is the one they actually end up running with in the guest OS. Once the initial boot state is attested to, the problem from a confidential computing POV is solved.
> 
> Attesting the ongoing running state of a VM to detect a malicious guest admin, or detect a compromise of the guest OS is a separate, but complementary problem, that may or may not be required for any given VM. This will certainly have some overlap in terms of technologies used, but will not necessarily require the same approach for all cases. Tracking ongoing mount table modifications post-boot is a massively more complex task given the widespread use of namespaces in Linux these days.

It seems to me the line being drawn here is extremely arbitrary, and doesn't make much sense to me. Why is there any practical difference between the first mount operation done by the gpt generator and any following one that happens immediately after? The end result is that the system is compromised. So knowing that the first client did the right thing really doesn't tell you much. You want to attest against _everything_ that the DM subsystem does - and to my understanding this is already mostly done and the IMA log should have all relevant events, including some details on the key for dm-crypt events.

**Comment by poettering** [comment]
@bluca how can IMA deliver mount measurements btw, if userspace issues them in undetermined order and in context of mount namespaces?

**Comment by bluca** [comment]
AFAIK it doesn't measure mounts, it measures DM events (dm-crypt/verity being set up, removed, etc)

**Comment by kraxel** [comment]
> It seems to me the line being drawn here is extremely arbitrary, and doesn't make much sense to me. Why is there any practical difference between the first mount operation done by the gpt generator and any following one that happens immediately after? 

Well, the gpt is not encrypted so it is easy to make the guest behave in specific ways by the host tampering with it.  It must be considered untrusted input.

For the root filesystem this is not the case, it can be protected by dm-verity or encryption or both.  So mounts triggered by files on the root filesystem (be that fstab or mount units or some script) is a completely different story.

**Comment by bluca** [comment]
> > It seems to me the line being drawn here is extremely arbitrary, and doesn't make much sense to me. Why is there any practical difference between the first mount operation done by the gpt generator and any following one that happens immediately after?
> 
> Well, the gpt is not encrypted so it is easy to make the guest behave in specific ways by the host tampering with it. It must be considered untrusted input.
> 
> For the root filesystem this is not the case, it can be protected by dm-verity or encryption or both. So mounts triggered by files on the root filesystem (be that fstab or mount units or some script) is a completely different story.

How do you know there aren't other disks that get mounted, by something else other than systemd, stored in exactly the same way?

**Comment by poettering** [comment]
> Well, the gpt is not encrypted so it is easy to make the guest behave in specific ways by the host tampering with it. It must be considered untrusted input.

Maybe we we should simply measure the GPT partition table when we use it? (doesn't uefi do that anyway somewhere?) that is sufficiently short and everything that gpt-auto discovery stuff makes its decisions on?

**Comment by kraxel** [comment]
> How do you know there aren't other disks that get mounted, by something else other than systemd, stored in exactly the same way?

That wouldn't happen out-of-the-blue.  It'll be triggered either by something on the root filesystem, or by some untrusted external input (say cloud-init).  Untrusted external input must either be ignored or measured.

**Comment by kraxel** [comment]
> Maybe we we should simply measure the GPT partition table when we use it? (doesn't uefi do that anyway somewhere?) that is sufficiently short and everything that gpt-auto discovery stuff makes its decisions on?

That idea just crossed my mind too, makes sense to me.

**Comment by poettering** [comment]
ok, so i get the sense you don't actually care about what is mounted or which luks volumes are attached. What you do care is that all data that flows into the decision what and how to mount/setup is measured – or comes from a trusted source –, right?

what is mounted and attached is currently sourced from various places:

1. /etc/fstab + /etc/crypttab – comes from trusted source
2. kernel cmdline – trusted + measured through sd-stub and secureboot if unified kernels are used
3. GPT partition table – which we could tackle if we'd measure the GPT part table early on that gpt-auto operates on
4. credentials passed into the OS – those picked from kernel cmdline and the ESP are already measured. Those from SMBIOS type 11 are not so far (or is SMBIOS measured as part of firmware stuff, anyway?
5. stuff from cloud-init/ignition – would be up to cloud-init/ignition
6. input from console – the "headless" option from console is too invasive i understand, so this would need something

what else?

**Comment by bluca** [comment]
> > How do you know there aren't other disks that get mounted, by something else other than systemd, stored in exactly the same way?
> 
> That wouldn't happen out-of-the-blue. It'll be triggered either by something on the root filesystem, or by some untrusted external input (say cloud-init). Untrusted external input must either be ignored or measured.

Well it has to be measured too. Hence my point: how do you do that, without doing this from the kernel side, which is where things actually happen?

Or to see it from the other side. Why focus on one of the many components that could send a request, rather than on the single component that actually implements all requests?

**Comment by kraxel** [comment]
> btw, systemd nowadays picks ups credentials (i.e. this stuff: https://systemd.io/CREDENTIALS/) from smbios type 11. Should systemd measure them if we pick them up from there? I presume so?

Good question.  The firmware loads that from qemu, and I remember plans to have the firmware measure everything it gets from qemu via fw_cfg.  Need to ask what the status/plans are.  Measuring this twice would be pointless ...

**Comment by berrange** [comment]
> I see, so it's a lot more about having measurements of what happened than about blocking what can happen, right?
> 
> I don't really grok what "confidential computing" really means. I mean, I know some keywords, but it appears to me there's a lot of handwaving and unclear semantics...
> 
> Is there a way to detect that I am running in a CC environment? Is there even a precise semantic definition of what this would even mean?

In a traditional VM environment you implicitly have to trust both the host hardware, and the host OS/hypervisor software stack and their administrator.

In a confidential computing VM environment, you establish a trust relationship with the host hardware. This involves communication with a remote attestation service and various cryptographic checks to establish integrity of the hardware, the guest firmware and kernel/initd. I won't get into details since that varies per CPU vendor, but the conceptual approach shares alot of heritage with SecureBoot and vTPMs. The result though, is that the VM admin can prove that their VM is running on hardware that has encrypted all guest RAM, CPU registers, page tables and various other state. 

Thus if the initial attestation is successful, you have eliminated the host OS/hypervisor software stack & administrator from being trusted parties for core VM state. If the host OS/admin is malicious, the most they can do is inflict denial of services attacks. They cannot see any data from your VM, except for memory pages that the guest OS has intentionally shared with the host by turning off encryption for that page.

While the VM CPU state and RAM is confidential, this still leaves all the emulated device backends as untrustworthy because they're executing in host OS context and the VM has to share memory with the host for these devices. Thus there needs to be encryption used for any device. For disks this is obviously LUKS/BitLocker/etc, while for NICs this is TLS or equivalent.  Other devices like video adapters, sound, keyboard, mouse, etc have no pratical way to use encryption so will be permanently untrustworthy and not desirable to use.

So we've proved the VM is confidential and the remote attestation service has given us keys to unlock a secure vTPM, which can in turn unlock the LUKS volume on the root filesystem. We "merely" need to establish that the LUKS volume that was unlocked is the one that actually matched the keys from the vTPM, and that the partitions mounted at time of 'switch root' all came from inside the LUKS volume.   From that point onwards we know anything executing is based on content present in our root filesystem. Verification of ongoing runtime state is upto whatever was installed in that root FS

**Comment by poettering** [comment]
> Good question. The firmware loads that from qemu, and I remember plans to have the firmware measure everything it gets from qemu via fw_cfg. Need to ask what the status/plans are. Measuring this twice would be pointless ...

I am more interested in the general case. i.e. qemu might measure this, but what about other hypervisors? And there apparently is interest supportng sbios type 11 in redfish, i.e. bare metal, so what there, can we expect this stuff to be measured already if set?

**Comment by berrange** [comment]
> btw, systemd nowadays picks ups credentials (i.e. this stuff: https://systemd.io/CREDENTIALS/) from smbios type 11. Should systemd measure them if we pick them up from there? I presume so?

SMBIOS data is fully visible to the host OS. So if any sensitive data is being provided to the guest via SMBIOS type 11, it needs to be encrypted prior to being given to the host OS for injection, and then decrypted by the guest OS when received. Non-sensitive data could be provided in clear text, but if it is used for any functional purpose it would require measuring.

**Comment by bluca** [comment]
> So we've proved the VM is confidential and the remote attestation service has given us keys to unlock a secure vTPM, which can in turn unlock the LUKS volume on the root filesystem. We "merely" need to establish that the LUKS volume that was unlocked is the one that actually matched the keys from the vTPM, and that the partitions mounted at time of 'switch root' all came from inside the LUKS volume. From that point onwards we know anything executing is based on content present in our root filesystem. Verification of ongoing runtime state is upto whatever was installed in that root FS

To further expand on my point above: by measuring the request, you actually are not proving that the LUKS volume that was unlocked is the one that actually matched the keys.
You are only proving that you _asked_ to use that volume. Only the kernel can attest on what actually _happens_.

**Comment by kraxel** [comment]
> efi variables that systemd reads (or are those measured anyway by firmware?)

Not needed I think.  In confidential VMs the initial state of the variable store is measured, and IIRC measuring all variable updates is standard uefi functionality.

**Comment by berrange** [comment]
> > efi variables that systemd reads (or are those measured anyway by firmware?)
> 
> Not needed I think. In confidential VMs the initial state of the variable store is measured, and IIRC measuring all variable updates is standard uefi functionality.

Confidential VMs won't have persistent variable storage anyway. The OVMF builds are done such that variables are non-persistent / stateless. Since we measure the known good firmware, we can be confident we've not received any efi variables with non-default values.

**Comment by berrange** [comment]
> > Well, the gpt is not encrypted so it is easy to make the guest behave in specific ways by the host tampering with it. It must be considered untrusted input.
> 
> Maybe we we should simply measure the GPT partition table when we use it? (doesn't uefi do that anyway somewhere?) that is sufficiently short and everything that gpt-auto discovery stuff makes its decisions on?

Right, that's probably actually sufficient to avoid needing to measure individual mounts. I believe we still need to measure how we unlocked any LUKS volume, to prove that we unlocked it using the key obtained from the trusted vTPM - ie to show someone didn't substitute a disk image with identical GPT table but with a different encrypted volume content unlocked in some other way

**Comment by kraxel** [comment]
> Confidential VMs won't have persistent variable storage anyway. The OVMF builds are done such that variables are non-persistent / stateless.

The initial state can be changed though.

> Since we measure the known good firmware, we can be confident we've not received any efi variables with non-default values.

It's a bit more complicated:
 * intel-tdx measures the code only at launch time, the initial variable store is measured by the firmware.
 * amd-sev measures the complete firmware image (containing both code and varstore) at launch.

But, yes, bottom line is variable store state is measured.

**Comment by bluca** [comment]
> > > Well, the gpt is not encrypted so it is easy to make the guest behave in specific ways by the host tampering with it. It must be considered untrusted input.
> > 
> > 
> > Maybe we we should simply measure the GPT partition table when we use it? (doesn't uefi do that anyway somewhere?) that is sufficiently short and everything that gpt-auto discovery stuff makes its decisions on?
> 
> Right, that's probably actually sufficient to avoid needing to measure individual mounts. I believe we still need to measure how we unlocked any LUKS volume, to prove that we unlocked it using the key obtained from the trusted vTPM - ie to show someone didn't substitute a disk image with identical GPT table but with a different encrypted volume content unlocked in some other way

Again, IMA already measures metadata on the key used to open a dm-crypt volume

**Comment by poettering** [comment]
> > efi variables that systemd reads (or are those measured anyway by firmware?)
> 
> Not needed I think. In confidential VMs the initial state of the variable store is measured, and IIRC measuring all variable updates is standard uefi functionality.

But i wonder what about bare metal and non-confidential VMs...

**Comment by poettering** [comment]
> Again, IMA already measures metadata on the key used to open a dm-crypt volume

IMA has no clue about the LUKS superblock has it? and also no clue which slot was actually used from userspace.

**Comment by bluca** [comment]
> > Again, IMA already measures metadata on the key used to open a dm-crypt volume
> 
> IMA has no clue about the LUKS superblock has it? and also no clue which slot was actually used from userspace.

It doesn't, but userspace doing the remote attestation verification should, no? Kernel as source of truth records state -> userspace verifies it all matches with expectations. To me this makes more sense than userspace measures its requests, and then verifies that its requests are the requests it wanted... ignoring what actually happens under the hood

**Comment by poettering** [comment]
> It doesn't, but userspace doing the remote attestation verification should, no? 

How would it know if the LUKS slot binding to some TPM was used to unlock the rootfs, or some other key, that was manually entered?

**Comment by poettering** [comment]
btw, this reminds me that in the TODO list there are some items about turning systemd-repart into something that cannot just adjust the GPT partition table, but also validate it. i.e. you can tell it to ensure that only partitions x, y, z exist, and that they have these flags and uuids and labels and whatnot. And if these rules don't apply halt the boot.

Thus, a trusted initrd could run repart that would not only add what is missing but also ensure the stuff that exists matches expectations. It would be a bit more useful than just measuring the GPT, since it would allow variables: for example as long as the home partition has the right type/label/flags it doesn't matter what size it has, as long as it is above some limit or so...

**Comment by berrange** [comment]
>     1. teach systemd-cryptsetup to measure what it does to the PCR optionally, and enable that by default in the initrd.
> 
>     2. add a service (or add to some existing service) that measures the GPT part table to some PCR.
> 
>     3. add a service that measures a barrier value to the PCR during the initrd→host transition and ensures mounts done after the transition will measure differently than those before.

 I was thinking there would be enough info from 1&2, but re-considering, an explicit barrier is more assuredly robust for identifying the transition point, so makes sense.


> btw, this reminds me that in the TODO list there are some items about turning systemd-repart into something that cannot just adjust the GPT partition table, but also validate it. i.e. you can tell it to ensure that only partitions x, y, z exist, and that they have these flags and uuids and labels and whatnot. And if these rules don't apply halt the boot.
> 
> Thus, a trusted initrd could run repart that would not only add what is missing but also ensure the stuff that exists matches expectations. It would be a bit more useful than just measuring the GPT, since it would allow variables: for example as long as the home partition has the right type/label/flags it doesn't matter what size it has, as long as it is above some limit or so...

That would be useful, as it is not uncommon to want to resize the rootfs partition (and subsequently grow its FS) when booting a a cloud VM.

**Comment by berrange** [comment]
> > Again, IMA already measures metadata on the key used to open a dm-crypt volume
> 
> IMA has no clue about the LUKS superblock has it? and also no clue which slot was actually used from userspace.

I took a look at https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/device-mapper/dm-ima.rst which shows a dm-crypt IMA example measurement:

```
 dm_version=4.45.0;
 name=crypt1,uuid=crypt_uuid1,major=253,minor=0,minor_count=1,num_targets=1;
 target_index=0,target_begin=0,target_len=1953125,target_name=crypt,target_version=1.23.0,
 allow_discards=y,same_cpu=n,submit_from_crypt_cpus=n,no_read_workqueue=n,no_write_workqueue=n,
 iv_large_sectors=n,cipher_string=aes-xts-plain64,key_size=32,key_parts=1,key_extra_size=0,key_mac_size=0;
```

If you trust the physical disk hardware attached to your machine (reasonable in the case of a physical machine in your possession) then this metadata looks sufficient to attest that you've opened the dm-crypt volume on the expected physical device with.  I don't see anything that ties specifically to the key associated with the volume though. The UUID is public information from the LUKS header so can be spoofed.  I was hoping for at least the hash of the master key to be present, to proof the identity of the LUKS payload. So, on its own, this doesn't appear to be able usable to detect where the disk was swapped out with a different disk. Seems like you would need to add in use of dm-verity to handle that scenario with IMA measurements, unless I've missed something ?

**Comment by bluca** [comment]
> > > Again, IMA already measures metadata on the key used to open a dm-crypt volume
> > 
> > 
> > IMA has no clue about the LUKS superblock has it? and also no clue which slot was actually used from userspace.
> 
> I took a look at https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/device-mapper/dm-ima.rst which shows a dm-crypt IMA example measurement:
> 
> ```
>  dm_version=4.45.0;
>  name=crypt1,uuid=crypt_uuid1,major=253,minor=0,minor_count=1,num_targets=1;
>  target_index=0,target_begin=0,target_len=1953125,target_name=crypt,target_version=1.23.0,
>  allow_discards=y,same_cpu=n,submit_from_crypt_cpus=n,no_read_workqueue=n,no_write_workqueue=n,
>  iv_large_sectors=n,cipher_string=aes-xts-plain64,key_size=32,key_parts=1,key_extra_size=0,key_mac_size=0;
> ```
> 
> If you trust the physical disk hardware attached to your machine (reasonable in the case of a physical machine in your possession) then this metadata looks sufficient to attest that you've opened the dm-crypt volume on the expected physical device with. I don't see anything that ties specifically to the key associated with the volume though. The UUID is public information from the LUKS header so can be spoofed. I was hoping for at least the hash of the master key to be present, to proof the identity of the LUKS payload. So, on its own, this doesn't appear to be able usable to detect where the disk was swapped out with a different disk. Seems like you would need to add in use of dm-verity to handle that scenario with IMA measurements, unless I've missed something ?

IIRC the roothash for verity is part of the measurement. A hash of the key or somesuch thing as you say is reasonable to have I think? So it should be a reasonable addition to the kernel, no?

**Comment by bluca** [comment]
> > It doesn't, but userspace doing the remote attestation verification should, no?
> 
> How would it know if the LUKS slot binding to some TPM was used to unlock the rootfs, or some other key, that was manually entered?

It would be the only one configured, no? And perhaps libcryptsetup should be enhanced to log information about the luks header it uses to do an activation?

**Comment by berrange** [comment]
> IIRC the roothash for verity is part of the measurement. A hash of the key or somesuch thing as you say is reasonable to have I think? So it should be a reasonable addition to the kernel, no?

Yes, I think it ought to be added.

> > How would it know if the LUKS slot binding to some TPM was used to unlock the rootfs, or some other key, that was manually entered?
> 
> It would be the only one configured, no? And perhaps libcryptsetup should be enhanced to log information about the luks header it uses to do an activation?

When first filing this systemd RFE, I wasn't sure that `cryptsetup` had enough info, since `systemd`'s `src/cryptsetup/cryptsetup.c` does alot of work to gather credentials, then passing it to `crypt_activate_by_xxx` APIs. I think I've convinced myself though that `libcryptsetup` should indeed have enough info. It would need to log information that covers much of the 'Keyslots' and 'Tokens' data reported by 'cryptsetup luksDump', along with the parameters to whichever `crypto_activate_by_xxx`  API is invoked.

The only possible gap I see is if the VM is given 2 vTPMs, you don't know which was used. I think that's probably a non-issue if the kernel IMA included a hash of the master key. Alternatively it might be that we want the kernel IMA log to be able to report details about every hardware device detected in the system more generally, to let user identify if the hypervisor attempts to trick the OS into using some unexpectedly exposed virtual hardware device.

**Comment by bluca** [comment]
> > > How would it know if the LUKS slot binding to some TPM was used to unlock the rootfs, or some other key, that was manually entered?
> > 
> > 
> > It would be the only one configured, no? And perhaps libcryptsetup should be enhanced to log information about the luks header it uses to do an activation?
> 
> When first filing this systemd RFE, I wasn't sure that `cryptsetup` had enough info, since `systemd`'s `src/cryptsetup/cryptsetup.c` does alot of work to gather credentials, then passing it to `crypt_activate_by_xxx` APIs. I think I've convinced myself though that `libcryptsetup` should indeed have enough info. It would need to log information that covers much of the 'Keyslots' and 'Tokens' data reported by 'cryptsetup luksDump', along with the parameters to whichever `crypto_activate_by_xxx` API is invoked.

Yeah it's cryptsetup that will have the visibility into the details of the luks header, _and_ the request that is passed to the kernel through the dm interface.
So the key thing there is, to ensure the log from cryptsetup can be matched to the corresponding dm-crypt log, and the hash of the volume key sounds like the right thing. That way you get the extra info from the full luks header which is only a userspace thing, but also get the confirmation that it's actually being used (and not just somebody spoofing libcryptsetup) by the kernel.

**Comment by DemiMarie** [comment]
> So I guess related to this we should probably revisit this, and maybe add an additional naming scheme /dev/disk/by-root-uuid/* or so which would only contain uuid symlinks and ptypeuuid symlinks for devices on the block device backing the rootfs itself. These links probably should even contain the "diskseq" in some form, to ensure they pinpoint exactly one specific device and no others. we'd then still need extra infra for /bin/mount to care about the diskseq though.

You need to check after opening, btw.  Otherwise there is a race condition.

**Comment by poettering** [comment]
@DemiMarie yes, see https://github.com/util-linux/util-linux/issues/1786 which would allow mounting a block device only if diskseq still matches

**Comment by poettering** [comment]
So after thinking a bit about all this my current approach for locking this down bit by bit is this:

1. Teach systemd's GPT disk image dissection logic a concept of "policies", that allow/disallow certain partitions to be around, and enforce requirements on them (i.e. must be verity, or must be encrypted, or must present, must be absent, is optional and so on). These then can be configured on the kernel command line for example. A the root disk doesn't match the policy we'll abort bootup. It would be systemd-gpt-auto-generator that enforces this. It would also generate the mount units based on this.

2. For verity volumes we'll add a new "tpm2-measure-pcr=" option to /etc/veritytab, that allows measuring the root hash on activation. systemd-gpt-auto-generator would automatically issue "tpm2-measure-pcr=13" for the root volume if verity is used (pcr 13 is the PCR we'll measure every "static" vendor supplied resources to that are "optional"/"modular"; which is unlike pcr 11 where we'll measure static vendor supplied resources that are mandatory to i.e. the kernel)

3. For encrypted volumes we'll add a new "tpm2-measure-pcr=" option to /etc/crypttab, that allows measuring the volume key on activation. systemd-gpt-auto-generator would automatically issue "tpm2-measure-pcr=15" for the root volume if encryption is used (pcr 15 is the PCR we'll measure local host ids into)

4. We'll also start measuring /etc/machine-id of the host from the initrd into PCR 15.

5. We'll measure the string "enter-initrd" during the kernel→initrd transition, as well as "leave-initrd" during initrd→host transition, as well as "ready" immediately before allowing regular users to login. At shutdown we'll measure "shutdown". All that into PCR 11 (where the kernel is measured into).

I am a bit unsure about measuring the selected LUKS enrollment slot (i.e. whether TPM was used or something else). I am not convinced this is necessary. To me it appears sufficient that if the volume key was successfully acquired then access to other resources should be granted too, regardless how we managed to acquire it.

I'd not bother with measuring the mounts we actually do, as the aforementioned dissection policy should be sufficient to ensure that only the stuff we want to mount is mounted.

So to summarize:

PCR 15 would be local instance/key material info. i.e. stuff that makes the local instance the local instance. PCR 13 would be the "optional" vendor stuff used to boot. i.e. which is not necessarily pinned by the chosen kernel, i.e. sysext images, chosen immutable root fs and so on. PCR 11 would the "mandatory" vendor stuff used to boot, i.e. the kernel image itself and its builtin initrd, and the boot phases we definitely will iterate through during the lifecycle of the OS

This way I think we can lock down everything we need: we'll measure all code, we'll measure the systems trusted handles to specific volumes (i.e. root hashes, volume keys) and to the instance itself (machine id). We'd do that all from the initrd or before. We can distinguish all relevant phases of the boot to make keys available only in some specific boot phases, and the phases are chosen so that they are immediately before we open the flood gates for additional external code/control. i.e. "enter-initrd" would open the floodgates for initrd sysexts, "leave-initrd" would open the floodgates to the host userspace, "ready" would open the floodgates to users logging in, and "shutdown" would be where we transition into the shutdown logic.

I implemented item nr 5 mostly here already, will post shortly. I think it's the most relevant to get done quickly, since it affects how we define PCR 11. Unlike the other PCRs PCR 11 is already fully hooked up with the upcoming release, hence I really want the boot phase stuff included in it, because adding it later would effectively break all code that started binding to PCR 11 already.

**Comment by poettering** [comment]
BTW since #24771 we'll measure the boot "phase" into PCR 11, at selected milestones. This enables us bind TPM secrets to certain phases of the boot, for example "unlock the disk secret only in the initrd, not after".

With #25006 there will be support for measuring:

1. LUKS volume keys (via the new `tpm2-measure-pcr=` option in `/etc/crypttab`, enabled by default for gpt-auto boots for the rootfs and `/var/`).
2. File system information, i.e. uuid, label, part-uuid (via the new `systemd.tpm2-measure-pcr=` mount option, also enabled by default for root fs and `/var/` if gpt-auto stuff is used)
3. `/etc/machine-id` (via a new `systemd-pcrmachine.service` service)

(when I say "by default" I mean "by default if booted via UKI", because otherwise we probably shouldn't interfere with pre-existing PCR assignments)

All these will be measured into PCR 15, in a defined order. After the early boot phase completes we thus guarantee a stable value in PCR 15 that clearly identifies the machine and can be used to bind secrets to the local system. By combining checks against PCR 11 and PCR 15 you can express "only on this kernel, in this boot phase, on this system".

Of my checklist above we are only missing the `tpm2-measure-pcr=` setting for `/etc/veritytab` I guess.

**Comment by berrange** [comment]
> I am a bit unsure about measuring the selected LUKS enrollment slot (i.e. whether TPM was used or something else). I am not convinced this is necessary. To me it appears sufficient that if the volume key was successfully acquired then access to other resources should be granted too, regardless how we managed to acquire it.

I'm still not comfortable with skipping this. From the POV of a confidential VM, the only mechanism that is safe to use is TPM based unlock. A disk image might be created with multiple keyslots present, one for TPM and one for recovery key. When booting that image on an in-house private cloud it would be fine to use either, but when booted on a public cloud in a confidential VM though, I'd like to be able to prove that it was unlocked using the TPM keyslot via a TPM measurement.  One could say you should build a different image for the public cloud vs private cloud usage, but my desire is to be able to build images that work in any scenario, where you just validate the TPM to determine whether it meets the level of trust desired for its execution environment.

> /etc/machine-id (via a new systemd-pcrmachine.service service)

I'm wondering if including machine-id could cause any complications with cloud images. Typically those are created with an empty `/etc/machine-id`, which is dynamically filled on first boot. IOW the `/etc/machine-id` value won't be known ahead of time, so if tieing secrets to the TPM measurements, it might be desirable to not include the machine ID. I guess the answer to this is probably to disable the `systemd-pcrmachine.service` unit if that's a problem scenario that is faced.

**Comment by poettering** [comment]
> > /etc/machine-id (via a new systemd-pcrmachine.service service)
> 
> I'm wondering if including machine-id could cause any complications with cloud images. Typically those are created with an empty `/etc/machine-id`, which is dynamically filled on first boot. IOW the `/etc/machine-id` value won't be known ahead of time, so if tieing secrets to the TPM measurements, it might be desirable to not include the machine ID. I guess the answer to this is probably to disable the `systemd-pcrmachine.service` unit if that's a problem scenario that is faced.

So in my view of the world the disk encryption key and the root fs is also generated at first boot, so all the stuff measured into PCR 15 is all in the same category: inherently local stuff, generated locally on first boot, and not provisioned ahead. Hence, yes, PCR 15 cannot be pre-determined, but that's kinda the point of it.

That said, I think for the CC cases there might be value in passing in a "seed" value into the system, and then deriving everything from that, to make it pre-calculatable. i.e. with a sufficiently large seed value we can generate the root disk encryption volume key, the machine ID, the root fs uuid and so on determenistically, by hashing that seed in counter mode or so.

in fact, this is actually what systemd-repart (our tool for adding/growing partitions at boot) already supports via its `--seed=` parameter (though this was done to ensure reproducible disk images). So maybe we should build on that, and not just add this scheme to systemd-repart for the stuff it generates, but also for the default machine ID generation we pass to PID 1 to use if none is set yet.

This seed value we'd then somehow have to propagate into the machine, and we should probably erase that before we transition from initrd to host. VMMs would only pass the seed in exactly once on first boot.

**Comment by poettering** [comment]
> I'm still not comfortable with skipping this. From the POV of a confidential VM, the only mechanism that is safe to use is TPM based unlock. A disk image might be created with multiple keyslots present, one for TPM and one for recovery key. When booting that image on an in-house private cloud it would be fine to use either, but when booted on a public cloud in a confidential VM though, I'd like to be able to prove that it was unlocked using the TPM keyslot via a TPM measurement. One could say you should build a different image for the public cloud vs private cloud usage, but my desire is to be able to build images that work in any scenario, where you just validate the TPM to determine whether it meets the level of trust desired for its execution environment.

so there are two ways to address this: allow unlocking by other means and measuring it. or simply not allowing it.

I am pretty sure in the CC case you want the latter, i.e. enforce some policy on what we'll automatically mount/activate from a disk image. And if the policy says "tpm only" then we'd honour that and refuse anything else.

I understand you would like to avoid having to configure that policy differently locally and in production. For that: is there a sane way to detect a CC environment? If we had that, we could simply pick slightly different good defaults for the policy. a more strict one that refuses interactivity for the CC case, and a less strict one elswhere.

I want that PCR 15 is something that is associated with a specific system and remains the same and predictable as long as the system stays the "same". For most cases this really means: it doesn't matter how the disk was unlocked, in all cases we end up with the same PCR 15, and that is a good thing. PCR 15 is hence something that people can directly lock secrets to that can only be unlocked locally.

Now, your case for measuring the used mechanism for unlocking has a different usecase: you care about remote attestation, i.e. the ability to reason after the fact if everything was done correctly. But that is a different usecase for PCRs afaics: it's more about using PCRs as validation backing source for analyzing log files then about having literal PCR values in place that policy can directly be bound to. Hence I am not saying we shouldn't measure this, but I am pretty sure it should be somewhere else than PCR 15.

There's also the question: it's not just about LUKS keys, is it? you probably want to cover all kind of local console stuff this way, including regular user log ins? But if that's the case aren't we just in this territory where we basically want to just have a TPM backed journal service? because we actually want to be able to sanely reason about pretty much *any* log event we see on the system?

If we see it that way: maybe we should teach journald native TPM support: every now and then extend a hashes of all log entries since the last time into some PCR? And then, in journalctl maybe transparently validate everything against the PCR? That would then include *everything* including information about what unlocked luks, who logged in on the console and so on. (of course, due to rotate you'll eventually lose the "anchor" for the validation, but if you pick up the data and store it centrally you can work around that.

In a way this is what IMA does, except that afaik IMA doesn't allow userspace to log arbitrary stuff to it...

**Comment by poettering** [comment]
So, everything from the list above is implemented, with the one exception of the slot/mechanism used to unlock the disk. I am nowadays actually quite sympathetic to the idea, for remote attestation, but it's not trivial to add, simply because we are out of PCRs. However, we can define our own PCRs via nvindexes in PCR mode. It's not particularly hard, and it can also be included in a tpm2 quote.

Anyway, let's close this issue for now, since the only thing left now is the chosen slot/mechanism. I created a new issue for that in #29877. Let's continue discussion there.
